### DISCUSSION TITLE
Different lenghs of theta parameter from sep.extract.

### DISCUSSION THREAD

**Discussion by igorkolesnikov13** [sep-developers/sep#119] [Q&A] [answered]
Hello guys, 
I'm trying to substitute sExtractor with SEP in my code and while testing I noticed strange behavior.
One of the parameters that my program needs is theta[2] from sep.extract.

What I have  noticed, is that the length of these parameters varies a lot. I have a testing set of 1000 galaxies, and half of them returned an error that position 2 does not exist. Without any error. While sextractor was able to extract THETA_IMAGE from all of them.

The lines that are being executed:
```
bkg = sep.Background(data)
data_sub = galaxyObject.originalImageData - bkg
objs, segmentedSep = sep.extract(data_sub, threshold, err=bkg.globalrms, segmentation_map=True, maskthresh=1.5)
galaxyObject.angle = objs['theta'][2]
```

Is there is a way to debug why SEP can't return theta[2] with some images?

Couple of problematic images: https://drive.google.com/file/d/1T3u6ZygScPHSCyUp2TLHJtsVklOH4J-0/view?usp=sharing

**Comment by kbarbary** [comment]
If `objs['theta'][2]` doesn't exist, I think this means that less than 3 objects were detected in the image (this would be the `THETA` parameter for the third object detected in the image).

When you say that the length of the parameters varies a lot, do you mean `len(objs)` varies from one image to another? When you say you have a testing set of 1000 galaxies, are they all on the same image? Or one galaxy per image?

**Reply by igorkolesnikov13** [reply]
Indeed, I have misinterpreted the documentation.
Now it makes sense.

In this case, the position (in `objs['theta][postion]`) will match segmentation_map index?

Anseweing your questions.
_When you say that the length of the parameters varies a lot, do you mean len(objs) varies from one image to another?_
No, I was meaning that some images did not have `objs['theta'][2]`.

_When you say you have a testing set of 1000 galaxies, are they all on the same image?_
No, each galaxy is in a separate image (positioned in the center).

_Or one galaxy per image?_
Exactly, one galaxy per image. But in some cases, it can contain secondary objects to be ignored. I need to consider only the main object (galaxy) in the center.

**Comment by kbarbary** [comment]
> In this case, the position (in objs['theta][postion]) will match segmentation_map index?

Almost. They're off by 1. From [the docs for `sep.extract`](https://sep.readthedocs.io/en/latest/api/sep.extract.html):

> segmap:ndarray, optional
Array of integers with same shape as data. Pixels not belonging to any object have value 0. All pixels belonging to the i-th object (e.g., objects[i]) have value i+1. Only returned if segmentation_map=True.

**Comment by kbarbary** [comment, marked as answer]
> Exactly, one galaxy per image. But in some cases, it can contain secondary objects to be ignored. I need to consider only the main object (galaxy) in the center.

I see. So the problem is that ideally `sep.extract` would return exactly one object (`len(objs) == 1`). However, you may have other detections in the image due to noise, or other galaxies you wish to ignore. If you know where on the image the object of interest is, I'd recommend selecting it from the array of objects based on position, like
```python
mask = (objs["x"] > xmin) & (objs["x"] < xmax) &  (objs["y"] > ymin) & (objs["y"] < ymax)
objects_in_box = objs[mask]
object_of_interest = objects_in_box[0]  # hopefully exactly one galaxy is in the (xmin, xmax), (ymin,ymax) box
```

**Reply by igorkolesnikov13** [reply]
Just perfect. I just tested it on a single image. Seems to be working flawlessly (returns single object in the center)! Later I will test on several thousands to ensure that is working fine.
Previously I was using `numpy.argmax(objs['npix'])` to identify the main galaxy (as it, normally, brightest and biggest object in the image). In any case, your suggestion is much better.
